pallets/interest-accrual/src/lib.rs

```diff
@@ -132,82 +150,78 @@ pub mod pallet {
 		DebtAdjustmentFailed,
 		/// Emits when the interest rate was not used
 		NoSuchRate,
+		/// Emits when a historic rate was asked for from the future
+		NotInPast,
 	}
 
-	// TODO: add permissionless extrinsic to update any rate
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let mut weight = T::DbWeight::get().reads_writes(1, 1);
+			let version = Pallet::<T>::storage_version();
+			if version < 1 {
+				weight += migration::v1::migrate::<T>();
+			}
+			StorageVersion::<T>::set(1);
+			weight
+		}
+
+		fn on_initialize(_: T::BlockNumber) -> Weight {
+			let mut weight = T::DbWeight::get().reads_writes(1, 1);
+			let then = LastUpdated::<T>::get();
+			LastUpdated::<T>::set(Self::now());
+			let delta = Self::now() - then;
+			let _bits = Moment::BITS - delta.leading_zeros();
+			Rate::<T>::translate(|per_sec, mut rate: RateDetailsOf<T>| {
+				let new_rate =
+					match Self::calculate_accumulated_rate(per_sec, rate.accumulated_rate, then) {
+						Ok(rate) => rate,
+						Err(e) => panic!("Failure of interest accrual for rate {:?}", per_sec),
```

Comment:
We obviously shouldn't panic here. My inclination is to just remove the rate if calculation fails - it's obviously "wrong" in some way (probably way too large) and any loans which depend on it will also likely be "wrong" and will need to be re-issued with the correct interest rate.

Removing rates here will cause any updates to those loans to fail with a `NoSuchRate` error, but of course before this change they would have failed with an overflow in calculating the updated rate anyway.